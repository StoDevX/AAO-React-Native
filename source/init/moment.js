// @flow
import moment from 'moment-timezone'

// These next values made possible by
// http://stackoverflow.com/questions/42513276

// Tell Moment to round up, so that Moment will always report the
// time until something else in a way that matches the clock time.
// That is, at 5:22pm, Moment will report "8 minutes until 5:30pm"
// until the clock ticks over to 5:23pm.
// If we do Math.floor, we go the other direction: it will report
// "7 minutes until 5:30pm" as soon as 5:22pm starts.
moment.relativeTimeRounding(Math.ceil)

// Tell Moment that a minute is 60 seconds, for reporting purposes.
// It defaults to 45 seconds.
moment.relativeTimeThreshold('s', 60)
