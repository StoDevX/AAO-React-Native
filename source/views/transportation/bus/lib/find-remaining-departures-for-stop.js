// @flow

import type {BusTimetableEntry, DepartureTimeList} from '../types'
import type {BusStateEnum} from './get-current-bus-iteration'

type Args = {|
  +stop: BusTimetableEntry,
  +busStatus: BusStateEnum,
  +departureIndex: null | number,
|}

export function findRemainingDeparturesForStop(args: Args): DepartureTimeList {
  const {stop, busStatus, departureIndex} = args

  if (!stop) {
    return []
  }

  switch (busStatus) {
    case 'before-start': {
      return stop.departures.slice(0)
    }

    case 'after-end': {
      return stop.departures.slice(-1)
    }

    default: {
      return departureIndex !== null
        ? stop.departures.slice(departureIndex)
        : stop.departures.slice(0)
    }
  }
}
