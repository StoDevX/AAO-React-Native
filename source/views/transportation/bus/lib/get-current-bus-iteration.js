// @flow

import type moment from 'moment'
import find from 'lodash/find'
import findLast from 'lodash/findLast'
import findLastIndex from 'lodash/findLastIndex'
import type {BusSchedule, DepartureTimeList} from '../types'

const isTruthy = x => Boolean(x)

export type BusStateEnum =
	| 'none'
	| 'before-start'
	| 'after-end'
	| 'running'
	| 'between-rounds'
type ReturnVal = {
	status: BusStateEnum,
	times: DepartureTimeList,
	index: null | number,
	nextStart: ?moment,
}

export function getCurrentBusIteration(
	schedule: BusSchedule,
	now: moment,
): ReturnVal {
	// If the schedule is empty
	if (schedule.times.length === 0) {
		return {status: 'none', times: [], index: null, nextStart: null}
	}

	// Handle "now" being before or after the bus runs for the day
	const veryFirst = find(schedule.times[0], isTruthy)
	const veryLast = findLast(schedule.times[schedule.times.length - 1], isTruthy)

	// Start off by handling another empty-schedule case
	if (!veryFirst || !veryLast) {
		return {status: 'none', times: [], index: null, nextStart: null}
	} else if (now.isBefore(veryFirst)) {
		return {
			status: 'before-start',
			times: [],
			index: null,
			nextStart: veryFirst,
		}
	} else if (now.isAfter(veryLast)) {
		return {status: 'after-end', times: [], index: null, nextStart: null}
	}

	// The meat of this function: find the furthest timeset that now is part of.
	// Because we operate on sets, instead of on one giant list of stops, we
	// use isSameOrAfter to account for the gaps between iterations.
	const index = findLastIndex(schedule.times, stopTimes => {
		const first = find(stopTimes, isTruthy)
		const last = findLast(stopTimes, isTruthy)

		// Handle the case where stopTimes is empty
		if (!first || !last) {
			return false
		}

		// The only happy case!
		if (now.isSameOrAfter(first)) {
			return true
		}

		// Otherwise, nope
		return false
	})

	// If we found something, yay!
	if (index !== -1) {
		const times = schedule.times[index]
		const nextTimes = schedule.times[index + 1] || []

		const lastStopTime = findLast(times, isTruthy)
		const nextStart = find(nextTimes, isTruthy)

		// Check if we're between two iterations
		if (lastStopTime && nextStart && now.isBetween(lastStopTime, nextStart)) {
			return {
				status: 'between-rounds',
				times: nextTimes,
				index: index + 1,
				nextStart,
			}
		}

		// If we're not, then return "running"
		return {status: 'running', times, index, nextStart}
	}

	// Last ditch effort: we're not running at all
	return {status: 'none', times: [], index: null, nextStart: null}
}
