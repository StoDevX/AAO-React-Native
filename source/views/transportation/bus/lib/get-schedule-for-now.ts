import type {BusSchedule, DayOfWeek} from '../types'
import type {Moment} from 'moment-timezone'

const allDaysOfWeek: DayOfWeek[] = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa']

export function getScheduleForNow(
	schedules: BusSchedule[],
	now: Moment,
): BusSchedule {
	// now.day returns 0-6, Sunday to Saturday
	let thisWeekday = allDaysOfWeek[now.day()]

	if (!thisWeekday) {
		throw new Error('Invalid weekday')
	}

	let schedule = schedules.find((instance) =>
		instance.days.includes(thisWeekday),
	)

	if (!schedule) {
		return {
			days: [thisWeekday],
			timetable: [],
			coordinates: {},
			stops: [],
			times: [],
		}
	}

	return schedule
}
