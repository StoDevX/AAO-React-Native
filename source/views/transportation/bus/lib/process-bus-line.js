// @flow

import type {
	UnprocessedBusLine,
	BusLine,
	UnprocessedBusSchedule,
	BusSchedule,
} from '../types'

import {parseTime} from './parse-time'
import type moment from 'moment'

export const processBusSchedule = (now: moment) => (
	scheduleData: UnprocessedBusSchedule,
): BusSchedule => {
	const times = scheduleData.times.map(timeList => timeList.map(parseTime(now)))

	const timetable = scheduleData.stops.map((stopName, i) => {
		let stop = {}
		stop.name = stopName

		if (stopName in scheduleData.coordinates) {
			stop.coordinates = scheduleData.coordinates[stopName]
		}

		stop.departures = times.map(timeList => timeList[i])

		return stop
	})

	return {
		days: scheduleData.days,
		coordinates: scheduleData.coordinates,
		stops: scheduleData.stops,
		times: times,
		timetable: timetable,
	}
}

export function processBusLine(
	lineData: UnprocessedBusLine,
	now: moment,
): BusLine {
	return {
		line: lineData.line,
		colors: lineData.colors,
		schedules: lineData.schedules.map(processBusSchedule(now)),
	}
}
