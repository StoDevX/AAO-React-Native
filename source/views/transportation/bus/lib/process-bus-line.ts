import type {
	BusLine,
	BusSchedule,
	UnprocessedBusLine,
	UnprocessedBusSchedule,
} from '../types'
import {BusTimetableEntry} from '../types'

import {parseTime} from './parse-time'
import type {Moment} from 'moment'

export const processBusSchedule =
	(now: Moment) =>
	(scheduleData: UnprocessedBusSchedule): BusSchedule => {
		let times = scheduleData.times.map((timeList) =>
			timeList.map(parseTime(now)),
		)

		let timetable = scheduleData.stops.map((stopName, i) => {
			let coordinates = scheduleData.coordinates[stopName]
			let departures = times.map((timeList) => timeList[i] ?? null)
			let stop: BusTimetableEntry = {name: stopName, departures, coordinates}
			return stop
		})

		return {
			days: scheduleData.days,
			coordinates: scheduleData.coordinates,
			stops: scheduleData.stops,
			times: times,
			timetable: timetable,
		}
	}

export function processBusLine(
	lineData: UnprocessedBusLine,
	now: Moment,
): BusLine {
	return {
		line: lineData.line,
		colors: lineData.colors,
		schedules: lineData.schedules.map(processBusSchedule(now)),
	}
}
