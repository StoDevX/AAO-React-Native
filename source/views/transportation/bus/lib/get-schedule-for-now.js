// @flow
import type {BusSchedule, DayOfWeek} from '../types'
import moment from 'moment-timezone'

const allDaysOfWeek: DayOfWeek[] = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa']

export function getScheduleForNow(
  schedules: BusSchedule[],
  now: moment,
): BusSchedule {
  // now.day returns 0-6, Sunday to Saturday
  const thisWeekday = allDaysOfWeek[now.day()]

  const schedule = schedules.find(schedule =>
    schedule.days.includes(thisWeekday),
  )

  if (!schedule) {
    return {
      days: [thisWeekday],
      timetable: [],
      coordinates: {},
      stops: [],
      times: [],
    }
  }

  return schedule
}
