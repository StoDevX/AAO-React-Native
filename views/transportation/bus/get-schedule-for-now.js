// @flow
import type {BusScheduleType} from './types'
import moment from 'moment-timezone'

const allDaysOfWeek = ['su', 'mo', 'tu', 'we', 'th', 'fr', 'sa']

export default function getScheduleForNow(
  schedules: BusScheduleType[],
  now: typeof moment
): BusScheduleType|void {
  const thisWeekday = now.day() // 0-6, Sunday to Saturday.

  return schedules.find(schedule => {
    return schedule.days
      .map(day => allDaysOfWeek.indexOf(day)) // turn days into numbers
      .includes(thisWeekday)
  })
}
