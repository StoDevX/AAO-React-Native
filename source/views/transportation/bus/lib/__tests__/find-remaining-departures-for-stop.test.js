// @flow
import {findRemainingDeparturesForStop} from '../find-remaining-departures-for-stop'

import {getScheduleForNow} from '../get-schedule-for-now'
import {getCurrentBusIteration} from '../get-current-bus-iteration'
import {processBusSchedule} from '../process-bus-line'
import {dayAndTime} from './moment.helper'

import type {UnprocessedBusSchedule, BusSchedule} from '../../types'

function buildBusSchedules(now): Array<BusSchedule> {
	// prettier-ignore
	const schedules: Array<UnprocessedBusSchedule> = [
    {
      days: ['Mo', 'Tu'],
      coordinates: {},
      stops: ['St. Olaf', 'Carleton', 'Third'],
      times: [['1:00pm',  '1:05pm',   '1:10pm'],
              ['2:00pm',  false,      '2:10pm'],
              ['3:00pm',  '3:05pm',   '3:10pm'],
      ],
    },
    {
      days: ['We'],
      coordinates: {},
      stops: ['St. Olaf', 'Carleton'],
      times: [['12:00pm', '12:05pm']],
    },
  ]
	return schedules.map(processBusSchedule(now))
}

const formatTime = m => (m ? m.format('h:mma') : null)

function makeSchedule(now) {
	const schedule = getScheduleForNow(buildBusSchedules(now), now)
	const {status, index} = getCurrentBusIteration(schedule, now)
	return {schedule, busStatus: status, departureIndex: index}
}

test('handles a time before the bus runs', () => {
	const now = dayAndTime('Mo 12:00pm')
	const {schedule, busStatus, departureIndex} = makeSchedule(now)
	const stop = schedule.timetable[0]
	const actual = findRemainingDeparturesForStop({
		stop,
		departureIndex,
		busStatus,
	})
	expect(actual.map(formatTime)).toMatchSnapshot()
})

test('handles a time right when the bus starts', () => {
	const now = dayAndTime('Mo 1:00pm')
	const {schedule, busStatus, departureIndex} = makeSchedule(now)
	const stop = schedule.timetable[0]
	const actual = findRemainingDeparturesForStop({
		stop,
		departureIndex,
		busStatus,
	})
	expect(actual.map(formatTime)).toMatchSnapshot()
})

test('handles a time between two stops', () => {
	const now = dayAndTime('Mo 1:01pm')
	const {schedule, busStatus, departureIndex} = makeSchedule(now)
	const stop = schedule.timetable[1]
	const actual = findRemainingDeparturesForStop({
		stop,
		departureIndex,
		busStatus,
	})
	expect(actual.map(formatTime)).toMatchSnapshot()
})

test('handles a time at the second stop', () => {
	const now = dayAndTime('Mo 1:05pm')
	const {schedule, busStatus, departureIndex} = makeSchedule(now)
	const stop = schedule.timetable[1]
	const actual = findRemainingDeparturesForStop({
		stop,
		departureIndex,
		busStatus,
	})
	expect(actual.map(formatTime)).toMatchSnapshot()
})

test('handles a time at the last stop', () => {
	const now = dayAndTime('Mo 1:10pm')
	const {schedule, busStatus, departureIndex} = makeSchedule(now)
	const stop = schedule.timetable[schedule.timetable.length - 1]
	const actual = findRemainingDeparturesForStop({
		stop,
		departureIndex,
		busStatus,
	})
	expect(actual.map(formatTime)).toMatchSnapshot()
})

test('handles a time between two iterations', () => {
	const now = dayAndTime('Mo 1:55pm')
	const {schedule, busStatus, departureIndex} = makeSchedule(now)
	const stop = schedule.timetable[0]
	const actual = findRemainingDeparturesForStop({
		stop,
		departureIndex,
		busStatus,
	})
	expect(actual.map(formatTime)).toMatchSnapshot()
})

test('handles the first stop of the second iteration', () => {
	const now = dayAndTime('Mo 2:00pm')
	const {schedule, busStatus, departureIndex} = makeSchedule(now)
	const stop = schedule.timetable[0]
	const actual = findRemainingDeparturesForStop({
		stop,
		departureIndex,
		busStatus,
	})
	expect(actual.map(formatTime)).toMatchSnapshot()
})

test('handles the first stop of the last iteration', () => {
	const now = dayAndTime('Mo 3:00pm')
	const {schedule, busStatus, departureIndex} = makeSchedule(now)
	const stop = schedule.timetable[0]
	const actual = findRemainingDeparturesForStop({
		stop,
		departureIndex,
		busStatus,
	})
	expect(actual.map(formatTime)).toMatchSnapshot()
})

test('handles the last stop of the last iteration', () => {
	const now = dayAndTime('Mo 3:10pm')
	const {schedule, busStatus, departureIndex} = makeSchedule(now)
	const stop = schedule.timetable[schedule.timetable.length - 1]
	const actual = findRemainingDeparturesForStop({
		stop,
		departureIndex,
		busStatus,
	})
	expect(actual.map(formatTime)).toMatchSnapshot()
})

test('handles a stop that will be skipped', () => {
	const now = dayAndTime('Mo 2:05pm')
	const {schedule, busStatus, departureIndex} = makeSchedule(now)
	const stop = schedule.timetable[1]
	const actual = findRemainingDeparturesForStop({
		stop,
		departureIndex,
		busStatus,
	})
	expect(actual.map(formatTime)).toMatchSnapshot()
})

test('handles a time after the bus runs', () => {
	const now = dayAndTime('Mo 11:00pm')
	const {schedule, busStatus, departureIndex} = makeSchedule(now)
	const stop = schedule.timetable[0]
	const actual = findRemainingDeparturesForStop({
		stop,
		departureIndex,
		busStatus,
	})
	expect(actual.map(formatTime)).toMatchSnapshot()
})

test('handles a time on another day', () => {
	const now = dayAndTime('We 12:00pm')
	const {schedule, busStatus, departureIndex} = makeSchedule(now)
	const stop = schedule.timetable[0]
	const actual = findRemainingDeparturesForStop({
		stop,
		departureIndex,
		busStatus,
	})
	expect(actual.map(formatTime)).toMatchSnapshot()
})

test('handles a time when the bus is not running', () => {
	const now = dayAndTime('Sa 1:00pm')
	const {schedule, busStatus, departureIndex} = makeSchedule(now)
	const stop = schedule.timetable[0]
	const actual = findRemainingDeparturesForStop({
		stop,
		departureIndex,
		busStatus,
	})
	expect(actual.map(formatTime)).toMatchSnapshot()
})
