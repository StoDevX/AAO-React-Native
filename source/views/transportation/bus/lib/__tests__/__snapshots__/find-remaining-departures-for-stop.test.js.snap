// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles a stop that will be skipped 1`] = `
Array [
  null,
  "3:05pm",
]
`;

exports[`handles a time after the bus runs 1`] = `
Array [
  "3:00pm",
]
`;

exports[`handles a time at the last stop 1`] = `
Array [
  "1:10pm",
  "2:10pm",
  "3:10pm",
]
`;

exports[`handles a time at the second stop 1`] = `
Array [
  "1:05pm",
  null,
  "3:05pm",
]
`;

exports[`handles a time before the bus runs 1`] = `
Array [
  "1:00pm",
  "2:00pm",
  "3:00pm",
]
`;

exports[`handles a time between two iterations 1`] = `
Array [
  "2:00pm",
  "3:00pm",
]
`;

exports[`handles a time between two stops 1`] = `
Array [
  "1:05pm",
  null,
  "3:05pm",
]
`;

exports[`handles a time on another day 1`] = `
Array [
  "12:00pm",
]
`;

exports[`handles a time right when the bus starts 1`] = `
Array [
  "1:00pm",
  "2:00pm",
  "3:00pm",
]
`;

exports[`handles a time when the bus is not running 1`] = `Array []`;

exports[`handles the first stop of the last iteration 1`] = `
Array [
  "3:00pm",
]
`;

exports[`handles the first stop of the second iteration 1`] = `
Array [
  "2:00pm",
  "3:00pm",
]
`;

exports[`handles the last stop of the last iteration 1`] = `
Array [
  "3:10pm",
]
`;
